library/oneview_volume.py

```diff
@@ -0,0 +1,319 @@
+    returned: on state 'present', but can be null.
+    type: complex
+'''
+
+VOLUME_CREATED = 'Volume added/created successfully.'
+VOLUME_UPDATED = 'Volume updated successfully.'
+VOLUME_DELETED = 'Volume removed/deleted successfully.'
+VOLUME_REPAIRED = 'Volume repaired successfully.'
+VOLUME_SNAPSHOT_CREATED = 'Volume snapshot created successfully.'
+VOLUME_SNAPSHOT_DELETED = 'Volume snapshot deleted successfully.'
+VOLUME_NOT_FOUND = 'Volume not found.'
+VOLUME_SNAPSHOT_NOT_FOUND = 'Snapshot not found.'
+VOLUME_ALREADY_ABSENT = 'Nothing to do.'
+VOLUME_NO_OPTIONS_PROVIDED = 'No options provided.'
+
+
+class VolumeModule(object):
+    argument_spec = dict(
+        config=dict(required=True, type='str'),
+        state=dict(
+            required=True,
+            choices=['present', 'absent', 'repaired', 'snapshot_created', 'snapshot_deleted']
+        ),
+        data=dict(required=True, type='dict'),
+        export_only=dict(required=False, type='bool'),
+    )
+
+    def __init__(self):
+        self.module = AnsibleModule(argument_spec=self.argument_spec, supports_check_mode=False)
+        self.oneview_client = OneViewClient.from_json_file(self.module.params['config'])
+
+    def run(self):
+        state = self.module.params['state']
+        data = self.module.params['data']
+
+        try:
+            if state == 'present':
+                self.__present(data)
+            elif state == 'absent':
+                export_only = self.module.params.get('export_only', False)
+                self.__absent(data, export_only)
+            elif state == 'repaired':
+                self.__repair(data)
+            elif state == 'snapshot_created':
+                self.__create_snapshot(data)
+            elif state == 'snapshot_deleted':
+                self.__delete_snapshot(data)
+
+        except Exception as exception:
+            self.module.fail_json(msg=exception.message)
+
+    def __present(self, data):
+        resource = self.__get_by_name(data)
+
+        if not resource:
+            self.__create(data)
+        else:
+            self.__update(data, resource)
+
+    def __absent(self, data, export_only):
+        resource = self.__get_by_name(data)
+
+        if resource:
+            self.oneview_client.volumes.delete(resource, export_only=export_only)
+            self.module.exit_json(changed=True,
+                                  msg=VOLUME_DELETED)
+        else:
+            self.module.exit_json(changed=False, msg=VOLUME_ALREADY_ABSENT)
+
+    def __create(self, data):
+        created_volume = self.oneview_client.volumes.create(data)
+
+        self.module.exit_json(changed=True,
+                              msg=VOLUME_CREATED,
+                              ansible_facts=dict(storage_volume=created_volume))
+
+    def __update(self, data, resource):
+        if 'newName' in data:
+            data['name'] = data['newName']
+            del data['newName']
+
```

Comment:
I think we decided to search for a resource with the name equals to the var 'newName' and fail when it already exist (name already in use, so you can not update)